project/yxuan0329/proposal.rst

```diff
@@ -0,0 +1,186 @@
+
+.. Analyze how your system takes input, produces results, provide interface, and
+.. performs any other operations.  Describe the system's work flow.  You may
+.. consider to use a flow chart but it is not required.  Specify the constraints
+.. assumed in your system.  Describe the modularization of the system.
+Architecture
+---------
+
+The library would include:
+
+- ``build_tree``: build a r-tree
+- ``search``: search for the available results in r-tree
+- ``insert``: insert a node in r-tree
+- ``delete``: delete a node in r-tree
+
+Datasets
+---------
+
+Randomly generate about 50-100 map data entry by Python for validating the 
+R-tree algorithm can works correctly. 
+
+
+API Description
+===============
+
+.. Show how your system can be programmed.  You are supposed to implement the
+.. system using both C++ and Python.  Describe how a user writes a script in the
+.. system.
+
+The users can include the library by linking in their project. 
+The C++ users can simply link by adding include file.
+
+``#include "r-tree-library.hpp"``
+
+The Python users can import the library by adding the script.
+
+``import r-tree-library``
+
+
+Engineering Infrastructure
+==========================
+
+.. Describe how you plan to put together the engineering system:
+
+.. 1. Automatic build system and how to build your program
+.. 2. Version control (show how you will use it)
+.. 3. Testing framework
+.. 4. Documentation
+
+.. Some of the above information should be included in the documentation in your
+.. software.
+
+.. You may use continuous integration, but it is not required.  If you use it,
+.. describe how it works in your code development.
+
+1. Automatic build system: CMake
+2. Version control: git
+3. Testing framework: Pytest
+4. Documentation: README.rst
+
+  Testing data source: random-generated testing datasets including location and the 
+  name of the place. 
+
+  The testing mesurement includes correctness and time performance.
+
+Schedule
+========
+
+.. Itemize the work to do and list them in a timeline.  Estimate the efforts of
+.. each item before the project starts.
+
+.. The schedule is expected to be adjusted during the development.  A schedule to
+.. accurately predict everything in the development is impossible and unnecessary.
+.. But the initial estimate is important, for the baseline becomes concrete when
+.. we adjust the plan.
+
+.. To start, you can list the to-do items on a weekly basis and assume the
+.. development of the project takes 8 weeks:
+
+* Planning phase (6 weeks from mm/dd to mm/dd):
+Week 1 (10/30)
+---------
+- Survey some reference about the data structure and implementation
+- Survey some existing R-tree libraries and run
+- Create a few testing dataset for testing the existing libraries
+
+Week 2 (11/06)
+---------
+- Create about 50-100 testing datasets for future use
+- Design the C++ classes architecture
+
+Week 3 (11/13)
+---------
+- Implement R-tree with C++: ``build_tree``, ``search``
```

Comment:
Tree building depends on insertion and deletion.  Search does not.  So you need to do build, insert, and delete together, and search is another bucket.

Reply:
Thanks, I've corrected them.